### PR TITLE
Adds progress reporting for all upload/download commands.

### PR DESCRIPTION
**Pull request by williexu**
Closes: https://github.com/Azure/azure-cli/issues/4744
Closes: https://github.com/Azure/azure-cli/issues/4543

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [y ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [ ] Each command and parameter has a meaningful description.
- [ ] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py [hunk 1]
```diff
@@ -548,6 +548,8 @@ def register_common_storage_account_options(context):
 for item in ['file', 'blob']:
     register_cli_argument('storage {} url'.format(item), 'protocol', help='Protocol to use.', default='https', **enum_choice_list(['http', 'https']))
     register_source_uri_arguments('storage {} copy start'.format(item))
+    for command in ['download', 'download-batch', 'upload', 'upload-batch']:
+        register_cli_argument('storage {} {}'.format(item, command), 'progress_callback', ignore_type, validator=add_progress_callback)
```
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py]: The essence of your change is using a parameter on the custom command and a validator to achieve a "dependency injection". This is the side effect of this approach. I do think we should look at knack to see if there is room to improve the overall mechanism of the core to allow dependency injection without explicit manipulate the arguments.

/cc @tjprescott

**Comment by derekbekoe** [comment]
@williexu @troydai What's the status of this PR?

**Comment by williexu** [comment]
I believe we were delaying this until after knack. A lot of customers do seem to want this feature though. @troydai what do you think?

**Comment by tjprescott** [comment]
Due to recent changes to the CLI core, you will need to rebase your changes against the latest in the dev branch.

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/4823
This is an **experimental preview** for @microsoft.com users.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py [hunk 2]
```diff
@@ -547,6 +545,11 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
         c.argument('source_container')
         c.argument('source_share')
 
+    for item in ['file', 'blob']:
+        for command in ['download', 'download-batch', 'upload', 'upload-batch']:
+            with self.argument_context('storage {} {}'.format(item, command)) as c:
+                c.argument('progress_callback', ignore_type, validator=add_progress_callback)
+
```
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py]: Move it into each command's own context.

**Comment by troydai** [comment]
change the --no-progress argument to a simple switch instead of three state flag. otherwise looks good.

**Comment by troydai** [comment]
Rebase and all good.

**Comment by troydai** [comment]
If the CI passes, please manually resolve the conflict in the history file and merge from the portal to save another round CI.

**Comment by williexu** [comment]
Canceled new CI build and merging since it passed last round.

**Pull request merged**